### PR TITLE
[DRAFT] Dapr content

### LINKED ISSUE #463: .NET aspire vs Dapr
### Type of issue

Missing information

### Description

1. Does .NET Aspire support production environments?
2. Where can .NET Aspire be applied effectively?
3. Provide more information about how .NET Aspire differs from Dapr.

### Page URL

https://learn.microsoft.com/en-us/dotnet/aspire/reference/aspire-faq

### Content source URL

https://github.com/dotnet/docs-aspire/blob/main/docs/reference/aspire-faq.yml

### Document Version Independent Id

54c22e7e-72c3-1685-4239-2d628f85e69f

### Article author

@IEvangelist

### LINKED ISSUE #534: [New article]: Aspire Dapr integration
### Proposed topic or title

Dapr integration

### Location in table of contents.

_No response_

### Reason for the article

Dapr and Aspire can work together, but there is a lot of overlap as well. Why would you combine these? Where does one end and the other begin? Where do they complement, or fight each other?

### Article abstract

Dapr and Aspire can work together, but there is a lot of overlap as well. Why would you combine these? Where does one end and the other begin? Where do they complement, or fight each other?

### Relevant searches

_No response_

---
[Associated WorkItem - 255430](https://dev.azure.com/msft-skilling/Content/_workitems/edit/255430)

### PR DESCRIPTION
**Pull request by IEvangelist**
## Summary

Dapr content

- Fixes #534
- Fixes #463

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/frameworks/dapr.md](https://github.com/dotnet/docs-aspire/blob/9664f1d81696f6b18873114f6d6f38c71b1f7e9d/docs/frameworks/dapr.md) | [Use Dapr with .NET Aspire](https://review.learn.microsoft.com/en-us/dotnet/aspire/frameworks/dapr?branch=pr-en-us-894) |

### CODE REVIEW

File: docs/frameworks/dapr.md [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+# Use Dapr with .NET Aspire
+
+[Distributed Application Runtime (Dapr)](https://docs.dapr.io/) offers developer APIs that run as a sidecar process and abstract away the common complexities of the underlying cloud platform. Dapr and .NET Aspire work together to improve your local development experience. By using Dapr with .NET Aspire, you can focus on writing and implementing .NET-based distributed applications instead of spending extra time with local onboarding.  
+
+In this guide, you'll learn how to take advantage of Dapr's abstraction and .NET Aspire's opinionated configuration of cloud technologies to build simple, portable, resilient, and secured microservices at-scale on Azure.
+
+## Prerequisites
+
+- TODO: Add prerequisites
```
Comment (philliphoff) [inline docs/frameworks/dapr.md:16]: I'd suggest making Dapr 1.13 (the latest release) a prerequisite (see: https://docs.dapr.io/getting-started/install-dapr-cli/), and also mention that the Dapr runtime should be initialized on the machine (see: https://docs.dapr.io/getting-started/install-dapr-selfhost/).

File: docs/frameworks/snippets/Dapr/Dapr.ServiceDefaults/Dapr.ServiceDefaults.csproj [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <ImplicitUsings>enable</ImplicitUsings>
+    <Nullable>enable</Nullable>
+    <IsAspireSharedProject>true</IsAspireSharedProject>
+  </PropertyGroup>
+
+  <ItemGroup>
+    <FrameworkReference Include="Microsoft.AspNetCore.App" />
+
+    <PackageReference Include="Microsoft.Extensions.Http.Resilience" Version="8.3.0" />
+    <PackageReference Include="Microsoft.Extensions.ServiceDiscovery" Version="8.0.0-preview.7.24251.11" />
```
Comment (IEvangelist) [inline docs/frameworks/snippets/Dapr/Dapr.ServiceDefaults/Dapr.ServiceDefaults.csproj:14]:
```suggestion
    <PackageReference Include="Microsoft.Extensions.ServiceDiscovery" Version="8.0.0" />
```

File: docs/frameworks/dapr.md [hunk 3]
```diff
@@ -0,0 +1,112 @@
...
+To install Dapr, see [Install the Dapr CLI](https://docs.dapr.io/getting-started/install-dapr-cli/). After installing the Dapr CLI, run the `dapr init` described in [Initialize Dapr in your local environment](https://docs.dapr.io/getting-started/install-dapr-selfhost/).
+
+## Get started
+
+To get started you need to add the Dapr hosting package to your app host project by installing the [Aspire.Hosting.Dapr](https://www.nuget.org/packages/Aspire.Hosting.Dapr) NuGet package.
+
+### [.NET CLI](#tab/dotnet-cli)
+
+```dotnetcli
+dotnet add package Aspire.Hosting.Dapr
```
Comment (philliphoff) [inline docs/frameworks/dapr.md:29]: At least prior to GA, you'll need the prerelease flag too, I think?

File: docs/frameworks/dapr.md [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+---
+
+For more information, see [dotnet add package](/dotnet/core/tools/dotnet-add-package) or [Manage package dependencies in .NET applications](/dotnet/core/tools/dependencies).
+
+The Dapr resource is added to the .NET Aspire distributed application builder using the `AddDapr()` method.
+An overload of the `AddDapr()` method that accepts Dapr options is available. For most applications, the default options will suffice.
+
+:::code language="csharp" source="snippets/Dapr/Dapr.AppHost/Program.cs" range="4":::
+
+Dapr uses the Sidecar pattern to run alongside your application. The Dapr sidecar is a lightweight, portable, and stateless HTTP server that listens for incoming HTTP requests from your application. The sidecar is responsible for managing the lifecycle of your application, including service discovery, configuration, and secrets management. To add a sidecar to a .NET Aspire resource by using the `WithDaprSidecar(string appId)` method. The `appId` parameter is the unique identifier for the Dapr application.
```
Comment (philliphoff) [inline docs/frameworks/dapr.md:48]: The Dapr sidecar doesn't start/stop your application, so I'm not sure I'd say that it manages its "lifecycle" so much as it serves as a conduit for interacting with other services and dependencies and abstracts the application from the specifics of those services and dependencies.

Comment (philliphoff) [inline docs/frameworks/dapr.md:48]: I wonder if we should lead with the `WithDaprSidecar()` overload as I believe that will infer the `appId` from the name given to the resource itself (which is probably what you'd want the `appId` to be as well).

File: docs/frameworks/dapr.md [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+
+:::code language="csharp" source="snippets/Dapr/Dapr.AppHost/Program.cs":::
+
+The .NET Aspire dashboard will shows the Dapr sidecar as a resource, with its status and logs.
+
+:::image type="content" source="media/aspire-dashboard-dapr-sidecar-resources.png" lightbox="media/aspire-dashboard-dapr-sidecar-resources.png" alt-text=".NET Aspire dashboard showing Dapr sidecar resources":::
+
+## Adding the Dapr SDK
+
+To use Dapr APIs from .NET Aspire resources you can use the [Dapr SDK for ASP.NET Core](https://www.nuget.org/packages/Dapr.AspNetCore/). The Dapr SDK provides a set of APIs to interact with Dapr sidecars.
```
Comment (philliphoff) [inline docs/frameworks/dapr.md:69]: A note that the `Dapr.AspNetCore` library is for when wanting to use some of the Dapr integration with ASP.NET (DI integration, registration of subscriptions, etc.); non-ASP.NET apps (e.g. console apps) can just use the `Dapr.Client` library if they just want to make calls through the Dapr sidecar.

File: docs/frameworks/dapr.md [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+
+The full `Program.cs` file for the frontend project shows the Dapr Client being added to the service builder and the `WeatherApiClient` class that uses the Dapr Client to call the backend service.
+
+:::code language="csharp" source="snippets/Dapr/Dapr.Web/Program.cs" highlight="15,17":::
+
+In, for example, a Blazor project the `WeatherApiClient` class can be injected into a component and used to call the backend service.
+
+:::code language="csharp" source="snippets/Dapr/Dapr.Web/Components/Pages/Weather.razor" highlight="5,47":::
+
+What actually happens when the Dapr SDK is used is that the Dapr sidecar is called over HTTP. The Dapr sidecar then forwards the request to the target service. The target service can be running in the same process as the Dapr sidecar or in a different process. The Dapr sidecar is responsible for service discovery and routing the request to the target service.
```
Comment (philliphoff) [inline docs/frameworks/dapr.md:104]: Note: the target service typically isn't running in the same process as the sidecar.  Or, rather, the component related to a service may run in the Dapr sidecar, but that will forward calls (e.g. pub-sub, output bindings, etc.) to some other service.

Comment (hhunter-ms) [reply docs/frameworks/dapr.md:104]: ✅

File: docs/frameworks/dapr.md (hunk 3 above)
Comment (hhunter-ms) [reply docs/frameworks/dapr.md:29]: @IEvangelist will this be published once this is GA?

File: docs/frameworks/dapr.md (hunk 4 above)
Comment (hhunter-ms) [reply docs/frameworks/dapr.md:48]: ✅

File: docs/frameworks/dapr.md (hunk 5 above)
Comment (hhunter-ms) [reply docs/frameworks/dapr.md:69]: ✅

File: docs/frameworks/dapr.md (hunk 4 above)
Comment (hhunter-ms) [reply docs/frameworks/dapr.md:48]: @NicoVermeir Would you be able to update your code to reflect this?

File: docs/frameworks/dapr.md [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+```xml
+<PackageReference Include="Aspire.Hosting.Dapr"
+                  Version="[SelectVersion]" />
+```
+
+---
+
+For more information, see [dotnet add package](/dotnet/core/tools/dotnet-add-package) or [Manage package dependencies in .NET applications](/dotnet/core/tools/dependencies).
+
+The Dapr resource is added to the .NET Aspire distributed application builder using the `AddDapr()` method.
```
[the remark this answers is not in the document]
Comment (hhunter-ms) [reply docs/frameworks/dapr.md:43]: @NicoVermeir

File: docs/frameworks/dapr.md (hunk 3 above)
Comment (IEvangelist) [reply docs/frameworks/dapr.md:29]: Yes, this goes live tomorrow morning, just before `//build`.

**Pull request merged**